### PR TITLE
mgr/dashboard: Fix CephFS chart

### PR DESCRIPTION
**Pull request by Devp00l**
There was a problem with the CephFS chart that causes the chart not to
show on initialization.

This PR also fixes that the auto refresh of the chart did not work.

Fixes: https://tracker.ceph.com/issues/41166
Fixes: https://tracker.ceph.com/issues/40096
Signed-off-by: Stephan Müller <smueller@suse.com>

I also added some tests and refactored the CephFS chart component.

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.ts [hunk 1]
```diff
@@ -22,136 +23,139 @@ export class CephfsChartComponent implements OnChanges, OnInit {
...
+      datasets: [
+        {
+          data: this.convert_timeseries(this.mdsCounter[this.lhsCounter])
+        },
+        {
+          data: this.delta_timeseries(this.mdsCounter[this.rhsCounter])
+        }
+      ] as ChartDataSets[]
+    });
   }
```
Comment (bk201) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.ts]: Sorry to ask here. I created [an issue](https://tracker.ceph.com/issues/40096) about the chart is not automatically refreshed as time advanced. I can make it work if I add the following code:

```typescript
    this.chart.datasets = [...this.chart.datasets];
```

Is this related to change detection of Angular? Thanks!

Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.ts]: Yes, some times, especially when using external components, you have to force a change detection for Angular to refresh the data.

Comment (Devp00l) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.ts]: @bk201 Thanks for the hint I will fix that :) Good spotting

Comment (Devp00l) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.ts]: @bk201 done :)

Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.ts]: Thanks for fixing this too!

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.spec.ts [hunk 2]
```diff
@@ -9,6 +9,12 @@ describe('CephfsChartComponent', () => {
   let component: CephfsChartComponent;
   let fixture: ComponentFixture<CephfsChartComponent>;
 
+  const createArray = () => {
```
Comment (p-na) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.spec.ts]: `createArray` always returns the same result. It does not need to be a function.

Comment (LenzGr) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.spec.ts]: So would it make more sense to create this array statically instead?

**Comment by Devp00l** [comment]
Addressed all comments

**Comment by Devp00l** [comment]
jenkins retest this please

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.ts [hunk 3]
```diff
@@ -22,136 +23,140 @@ export class CephfsChartComponent implements OnChanges, OnInit {
...
+        intersect: false,
+        position: 'nearest',
+        callbacks: {
+          // Pick the Unix timestamp of the first tooltip item.
+          title: function(tooltipItems, data): string {
+            let ts = 0;
+            if (tooltipItems.length > 0) {
+              const item = tooltipItems[0];
+              const point = data.datasets[item.datasetIndex].data[item.index] as ChartPoint;
+              ts = point.x as number;
```
Comment (p-na) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.ts]: If `point` is of type `ChartPoint`, does `point.x` really need the `as number` addition?

Comment (Devp00l) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.ts]: As `x` of `CharPoint` can be `number`, `string` or `Date`, `as number` is needed in order to assign it to the variable ts which holds a number.

Before the use of types here the code looked like this:
```ts
              let ts = 0;
              if (tooltipItems.length > 0) {
                const item = tooltipItems[0];
                ts = data.datasets[item.datasetIndex].data[item.index].x;
              }
              return ts;
```
But that will produce compile errors if types are in use.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.ts [hunk 4]
```diff
@@ -22,136 +23,140 @@ export class CephfsChartComponent implements OnChanges, OnInit {
...
+        ]
+      },
+      tooltips: {
+        enabled: false,
+        mode: 'index',
+        intersect: false,
+        position: 'nearest',
+        callbacks: {
+          // Pick the Unix timestamp of the first tooltip item.
+          title: function(tooltipItems, data): string {
```
Comment (p-na) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.ts]: Would this also work with an arrow function? I think my question is, is there a particular reason why this is not an arrow function?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.spec.ts (hunk 2 above)
Comment (p-na) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-chart/cephfs-chart.component.spec.ts]: > So would it make more sense to create this array statically instead?

I'd say so. Currently, the function is called two times, returning the same result. That's not much, but image we'd do it all the time. I'd prefer to have it as an array and pass that directly to reduce clutter. If there's a need to return variations of data, the function could easily be reimplemented and would then serve an IMO meaningful purpose.

**Comment by Devp00l** [comment]
Addressed all comments

**Comment by Devp00l** [comment]
jenkins test make check arm64

**Comment by Devp00l** [comment]
jenkins retest this please

**Comment by LenzGr** [comment]
retest this please

**Pull request merged**